pkg/controller/zerocapacity/zero_controller.go

```diff
@@ -0,0 +1,470 @@
+	reqLogger.Info("Reconciling " + resource)
+	defer reqLogger.Info("----- End Reconciling " + resource)
+
+	instance, err := reconciler.ResourceInstance(request.NamespacedName, z)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, fmt.Errorf("Unable to fetch %s CR '%s': %w", resource, request.Name, err)
+	}
+
+	phase := instance.Phase()
+	switch phase {
+	case "":
+		return reconcile.Result{}, instance.UpdatePhase(ZeroInitializing)
+	case ZeroInitializing:
+		return z.initializeResources(request, instance)
+	case ZeroSucceeded, ZeroFailed:
+		return z.cleanupResources(request)
+	}
+
+	infinispan := &v1.Infinispan{}
+	clusterName := instance.Cluster()
+	clusterObjKey := types.NamespacedName{
+		Namespace: namespace,
+		Name:      clusterName,
+	}
+	if err := z.Get(context.Background(), clusterObjKey, infinispan); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, fmt.Errorf("CR '%s' not found: %w", clusterName, err)
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, fmt.Errorf("Unable to fetch CR '%s': %w", clusterName, err)
+	}
+
+	user := consts.DefaultOperatorUser
+	pass, err := users.PasswordFromSecret(user, infinispan.GetSecretName(), namespace, z.Kube)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	httpConfig := ispnclient.HttpConfig{
+		Username:  user,
+		Password:  pass,
+		Namespace: namespace,
+		Protocol:  strings.ToLower(string(infinispan.GetEndpointScheme())),
+	}
+	httpClient := curl.New(httpConfig, z.Kube)
+
+	if phase == ZeroInitialized {
+		return z.execute(httpClient, request, instance)
+	}
+	// Phase must be ZeroRunning, so wait for execution to complete
+	return z.waitForExecutionToComplete(httpClient, request, instance)
+}
+
+func (z *Controller) initializeResources(request reconcile.Request, instance Resource) (reconcile.Result, error) {
+	ctx := context.Background()
+	name := request.Name
+	namespace := request.Namespace
+	clusterName := instance.Cluster()
+	clusterKey := types.NamespacedName{
+		Namespace: namespace,
+		Name:      clusterName,
+	}
+
+	infinispan := &v1.Infinispan{}
+	if err := z.Client.Get(ctx, clusterKey, infinispan); err != nil {
+		retErr := fmt.Errorf("Unable to load Infinispan Cluster '%s': %w", clusterName, err)
+		if errors.IsNotFound(err) {
+			return reconcile.Result{RequeueAfter: consts.DefaultWaitOnCluster}, retErr
+		}
+		return reconcile.Result{}, retErr
+	}
+
+	if err := ensureClusterStability(infinispan); err != nil {
+		return reconcile.Result{}, fmt.Errorf("Infinispan not stable: %w", err)
+	}
+
+	clusterKey.Name = infinispan.Status.StatefulSetName
```

Comment:
@ryanemerson the reason of the previous error can be that value of the `infinispan.Status.StatefulSetName` you handled is empty, why infinispan_controller not yet set it. Logic need to be changed
You are validating `clusterKey.Name` but prints in error `clusterName` that's worst practice

Reply:
I think it's easiest just to use the name of the already retrieved `Infinispan` CR as the statefulset is always created with that name.

Reply:
Yes, sure. As now `Status.statefulSetName` is optional with this commit
https://github.com/infinispan/infinispan-operator/commit/c0341a5660328b58f87ba0a2f18b05e191afd33c